scheduler/pkg/envoy/xdscache/seldoncache.go

```diff
@@ -56,24 +68,115 @@ type PipelineGatewayDetails struct {
-		permanentClusters:      make([]types.Resource, permanentClusterCount),
+func NewSeldonXDSCache(logger logrus.FieldLogger, pipelineGatewayDetails *PipelineGatewayDetails) (*SeldonXDSCache, error) {
+	xdsCache := &SeldonXDSCache{
 		Clusters:               make(map[string]Cluster),
-		Routes:                 make(map[string]Route),
 		Pipelines:              make(map[string]PipelineRoute),
-		Secrets:                make(map[string]Secret),
-		PipelineGatewayDetails: pipelineGatewayDetails,
+		Routes:                 make(map[string]Route),
+		clustersToAdd:          make(map[string]bool),
+		clustersToRemove:       make(map[string]bool),
+		pipelineGatewayDetails: pipelineGatewayDetails,
+		secrets:                make(map[string]Secret),
 		logger:                 logger.WithField("source", "SeldonXDSCache"),
+		snapshotVersion:        rand.Int63n(1000),
+	}
+	err := xdsCache.init()
+	if err != nil {
+		return nil, err
+	}
+	return xdsCache, nil
+}
+
+func (xds *SeldonXDSCache) CreateWatch(req *cache.Request, stream stream.StreamState, responseChan chan cache.Response) (cancel func()) {
+	return xds.muxCache.CreateWatch(req, stream, responseChan)
+}
+
+func (xds *SeldonXDSCache) CreateDeltaWatch(req *cache.DeltaRequest, stream stream.StreamState, responseChan chan cache.DeltaResponse) (cancel func()) {
+	return xds.muxCache.CreateDeltaWatch(req, stream, responseChan)
+}
+
+func (xds *SeldonXDSCache) Fetch(ctx context.Context, req *cache.Request) (cache.Response, error) {
+	return xds.muxCache.Fetch(ctx, req)
+}
+
+func (xds *SeldonXDSCache) newSnapshotVersion() string {
+	// Reset the snapshotVersion if it ever hits max size.
+	if xds.snapshotVersion == math.MaxInt64 {
+		xds.snapshotVersion = 0
 	}
+
+	// Increment the snapshot version & return as string.
+	xds.snapshotVersion++
+	return strconv.FormatInt(xds.snapshotVersion, 10)
 }
 
-func (xds *SeldonXDSCache) SetupTLS() error {
+func (xds *SeldonXDSCache) init() error {
+	linearLogger := xds.logger.WithField("source", "LinearCache")
+	snapshotLogger := xds.logger.WithField("source", "SnapshotCache")
+
+	secretTypeURL := resource.SecretType
+	sds := cache.NewLinearCache(secretTypeURL, cache.WithLogger(linearLogger))
+	xds.sds = sds
+
+	listenerTypeURL := resource.ListenerType
+	lds := cache.NewLinearCache(listenerTypeURL, cache.WithLogger(linearLogger))
+	xds.lds = lds
+
+	clusterTypeURL := resource.ClusterType
+	cds := cache.NewLinearCache(clusterTypeURL, cache.WithLogger(linearLogger))
+	xds.cds = cds
+
+	snapshotCache := cache.NewSnapshotCache(true, cache.IDHash{}, snapshotLogger)
+	xds.snapshotCache = snapshotCache
+
+	classify := func(typeUrl string) string {
+		switch typeUrl {
+		case secretTypeURL:
+			return secretTypeURL
+		case listenerTypeURL:
+			return listenerTypeURL
+		case clusterTypeURL:
+			return clusterTypeURL
+		default:
+			return snapshotType
+		}
+	}
+
+	xds.muxCache = &cache.MuxCache{
+		Classify: func(req *cache.Request) string {
```

Comment:
In practice `Classify` is not used as we use delta xDS? if so add a note.

Reply:
Classify is used the `muxCache` to determine which underlying cache to use when responding to a request.